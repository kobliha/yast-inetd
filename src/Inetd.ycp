/**
 * File:	modules/Inetd.ycp
 * Package:	Configuration of inetd
 * Summary:	Data for configuration of inetd, input and output functions.
 * Authors:	Petr Hadraba <phadraba@suse.cz>
 *
 * $Id$
 *
 * Representation of the configuration of inetd.
 * Input and output routines.
 */

{

    /**
     * New SCR interface:
     * A service map looks like this:
     * $[
     *   "service": string	// * different from equally named field above
     *   "rpc_version": string
     *   "socket_type": string
     *   "protocol": string
     *   "wait": boolean
     *   "max": integer		// inetd only
     *   "user": string		// *
     *   "group": string
     *   "server": string
     *   "server_args": string
     *   "comment": string	// possibly multiline, without #
     ** bookkeeping fields:
     *   "iid": string		// internal id, use as table `id
     *      Iid is necessary because there may be multiple variants
     *      of the same service.
     *   "changed": boolean	// when writing, unchanged services are ignored
     *                          //   new services (created) must be set as changed
     *                          //   see changeLine() and addLine() for more details
     *   "deleted": boolean	// when deleting, this is set to TRUE and changed
     *                          // must be set too (see deleteLine())
     * ]
     * When handling existing maps, take care to preserve any other fields
     * that may be present!
     *
     * path netd = .whatever.inetd or .whatever.xinetd;
     *
     * SCR::Read (netd + .services) -> list of service maps
     *   There should be a warning if inetd and xinetd are both installed
//     * SCR::Read (netd + .services + service_name) -> service map (or empty)
//     * SCR::Read (netd + .services + service_name + field) -> (any) field
     * SCR::Write (netd, nil) -> boolean
     *   Commits changes performed by other writes (see below) to disk
     * // Hm, maybe we should check that vaid fields are passed?
     * SCR::Write (netd + .services, list of service maps) -> true
     * SCR::Write (netd + .services + service_name, service map) -> true
     * SCR::Write (netd + .services + service_name + field, any) -> true
     *   This last one should be useful for autoyast - changing just
     *   the "enabled" field.
     */

module "Inetd";
textdomain "inetd";

import "Progress";
import "Report";
import "Summary";

include "inetd/routines.ycp";
include "inetd/default_conf_inetd.ycp";
include "inetd/default_conf_xinetd.ycp";

/**
 * Abort function
 * return boolean return true if abort
 */
global block AbortFunction = nil;

/**
 * Data was modified?
 */
global boolean modified = false;

global boolean proposal_valid = false;
global boolean write_only = false;

/**
 * Data was modified?
 * @return true if modified
 */
global define boolean Modified() ``{
    //y2debug("modified=%1",modified);
    return modified;
};

/**
 * hold informations about installed services (rpm -q exit code).
 * 0 means installed, anything else means no or error
 */
global integer inetd_installed   = -1;
global integer xinetd_installed  = -1;

/**
 * What service is ready for configuration?
 * `none   means no available --- this is for AutoInstallation
 * `both   means the selection dialog will be displayed (only one service can run)
 * `inetd  inetd will be configured
 * `xinetd xinetd will be configured
 */
global symbol netd_service       = `none;

/**
 * What service was configured?
 * Same values as netd_service except `both
 */
global symbol configured_service = `none;

/**
 * These variables hold service's configuration
 */
global list inetd_conf           = [];
global list xinetd_conf          = [];

/**
 * Holds configuration of selected service.
 */
global list netd_conf            = [];

/**
 * Is inetd (or xinetd) running?
 */
global integer inetd_status      = -1;
global integer xinetd_status     = -1;

/**
 * Status of edited service (see WhatToConfigureDialog() in dialogs.ycp)
 */
global integer netd_status       = -1;

/**
 * Read all inetd settings
 * @return true on success
 */
global define boolean Read() ``{

    /* Inetd read dialog caption */
    string caption = _("Initializing inetd configuration");

    integer steps = 1;

    integer sl = 0;
    sleep(sl);

    Progress::New( caption, " ", steps, [
	    _("Read the configuration"),
	], [
	    _("Reading the configuration..."),
	    _("Finished")
	],
	""
    );

    integer read_status = 0;
    // read database
    if(Abort()) return false;

    if(netd_service == `none)
	return false;

    if(netd_service == `both) {
	inetd_conf  = SCR::Read(.etc.inetd_conf.services);
	xinetd_conf = SCR::Read(.etc.xinetd_conf.services);
    }
    else if(netd_service == `inetd) {
	inetd_conf = SCR::Read(.etc.inetd_conf.services);
    }
    else if(netd_service == `xinetd) {
	xinetd_conf = SCR::Read(.etc.xinetd_conf.services);
    }
    // in future: catch error
    if(false) Report::Error(_("Can not read the database1!"));
    sleep(sl);

    if(Abort()) return false;
    ProgressNextStage(_("Finished"));
    sleep(sl);

    if(Abort()) return false;
    modified = false;
    Progress::Finish();
    return true;
}

/**
 * Write all inetd settings
 * @return true on success
 */
global define boolean Write() ``{

    /* Inetd read dialog caption */
    string caption = _("Saving inetd configuration");

    integer steps = 1;

    integer sl = 0;
    sleep(sl);

    Progress::New(caption, " ", steps, [
	    _("Write the settings"),
	], [
	    _("Writing the settings..."),
	    _("Finished")
	],
	""
    );

    y2milestone("Calling write:\n");
    // write settings
    if(Abort()) return false;
//    y2milestone("%1, %2", configured_service, netd_status);
    if (write_only) {
	if (configured_service == `inetd) {
	    if (netd_status == 0) {
		Runlevel::ServiceAdjust("inetd", "enable");
	    }
	    else {
		Runlevel::ServiceAdjust("inetd", "disable");
	    }
	    Runlevel::ServiceAdjust("xinetd", "disable");
	    SCR::Write(.etc.inetd_conf.services, netd_conf);
	}
	else if (configured_service == `xinetd) {
	    if (netd_status == 0) {
		Runlevel::ServiceAdjust("xinetd", "enable");
	    }
	    else {
		Runlevel::ServiceAdjust("xinetd", "disable");
	    }
	    Runlevel::ServiceAdjust("inetd", "disable");
	    SCR::Write(.etc.xinetd_conf.services, netd_conf);
	}
    }
    else {
    if (netd_status != 0) {
	y2milestone("Nothing changed  --- installed services will be stopped");
	if (inetd_installed == 0) {
	    y2milestone("inetd installed  --- stopping and diabling service");
	    Runlevel::RunInitScript("inetd", "stop");
	    Runlevel::ServiceAdjust("inetd", "disable");
	}
	else {
	    y2milestone("inetd not installed --- nothing done");
	}
	if (xinetd_installed == 0) {
	    y2milestone("xinetd installed --- stopping and disabling service");
	    Runlevel::RunInitScript("xinetd", "stop");
	    Runlevel::ServiceAdjust("xinetd", "disable");
	}
	else {
	    y2milestone("xinetd not installed --- nothing done");
	}
    }
    // custom
    else {
	// configured service is inetd
	if (configured_service == `inetd) {
	    // xinetd must be stopped - if installed
	    if (xinetd_installed == 0) {
		y2milestone("xinetd installed  --- stopping and disabling service");
		Runlevel::RunInitScript("xinetd", "stop");
		Runlevel::ServiceAdjust("xinetd", "disable");
	    }
	    else {
		y2milestone("xinetd not installed --- nothing done");
	    }
	    y2milestone("writing inetd configuration");
	    SCR::Write(.etc.inetd_conf.services, netd_conf);
	    // So, Restart service
	    // if inetd is running - only reload
	    if (inetd_status == 0) {
		y2milestone("inetd was running --- calling force-reload");
		Runlevel::RunInitScript("inetd", "force-reload");
	    }
	    // if stopped - enable and start
	    else {
		y2milestone("inetd was stopped --- enabling and starting service");
		Runlevel::ServiceAdjust("inetd", "enable");
		Runlevel::RunInitScript("inetd", "start");
	    }
	}
	// configured service is xinetd
	else {
	    // inetd must be stopped - if installed
	    if (inetd_installed == 0) {
		y2milestone("inetd installed     --- stopping and disabling service");
		Runlevel::RunInitScript("inetd", "stop");
		Runlevel::ServiceAdjust("inetd", "disable");
	    }
	    else {
		y2milestone("inetd not installed --- nothing done");
	    }
	    y2milestone("writing xinetd cnfiguration");
	    SCR::Write(.etc.xinetd_conf.services, netd_conf);
	    // So, Restart service
	    // if xinetd is running - only reload
	    if (inetd_status == 0) {
		y2milestone("xinetd was running --- calling force-reload");
		Runlevel::RunInitScript("xinetd", "force-reload");
	    }
	    // if stopped - enable and start
	    else {
		y2milestone("xinetd was stopped --- enabling and starting service");
		Runlevel::ServiceAdjust("xinetd", "enable");
		Runlevel::RunInitScript("xinetd", "start");
	    }
	}
    }
    }

    y2milestone("Writting done\n");

    // in future: catch errors
    if(false) Report::Error (_("Can not write settings!"));
    sleep(sl);

    if(Abort()) return false;
    ProgressNextStage(_("Finished"));
    sleep(sl);

    if(Abort()) return false;
    Progress::Finish();
    return true;
}

/**
 * Only Write settings
 * @return boolean True on success
 */
global define boolean WriteOnly () ``{
    write_only = true;
    return Write();
}

/**
 * Get all inetd settings from the first parameter
 * (For use by autoinstallation.)
 * @param settings The YCP structure to be imported.
 * @return boolean True on success
 */
global define boolean Import (map settings) ``{
//    sleep(3000);
    /*
     * if map contains "netd_service", whole configuration is filled from this map.
     * if "netd_service" is not found, configuration is generated from default values
     * stored if default_conf_*.ycp.
     */
    if (settings["netd_service"]:`none == `inetd) {
	inetd_conf    = settings["netd_conf"]:[];
        inetd_status  = 0;
	netd_service  = `inetd;
	netd_status   = settings["netd_status"]:-1;
	netd_conf     = inetd_conf;
    }
    else if (settings["netd_service"]:`none == `xinetd) {
	xinetd_conf   = settings["netd_conf"]:[];
        xinetd_status = 0;
	netd_service  = `xinetd;
	netd_status   = settings["netd_status"]:-1;
	netd_conf     = xinetd_conf;
    }
    else if (settings["netd_service"]:`none == `none) {
	inetd_conf    = default_conf_inetd["netd_conf"]:[];
	xinetd_conf   = default_conf_xinetd["netd_conf"]:[];
	inetd_status  = 0;
	xinetd_status = 0;
	netd_service  = `both;
    }
    configured_service = netd_service;
    return true;
}

/**
 * Dump the inetd settings to a single map
 * (For use by autoinstallation.)
 * @return map Dumped settings (later acceptable by Import ())
 */
global define map Export () ``{
//    sleep(3000);
    map config = $[];
    config = add(config, "netd_conf",    netd_conf);
    config = add(config, "netd_service", configured_service);
    config = add(config, "netd_status",  netd_status);
    y2milestone("%1", config);
    return config;
//    return $[];
}

/**
 * Create a textual summary and a list of unconfigured cards
 * // @param split split configured and unconfigured?
 * @return summary of the current configuration
 */
global define list Summary() ``{
    // TODO FIXME: your code here...
    return [ _("Configuration summary ..."), [] ];
}

/**
 * Create an overview table with all configured cards
 * @return table items
 */
global define list Overview() ``{
    // TODO FIXME: your code here...
    return [];
}

/**
 * delete line in netd_conf
 * @param line_number "iid" geted from table's item ID
 * @return map If success returns updated line
 */
global define deleteLine(any line_number) ``{
    // delete
    map current_line = select(filter(`line, netd_conf,
	``(lookup(line, "iid", "0") == line_number)), 0, $[]);
    // set "deleted" flag to true
    current_line = add(current_line, "changed", true);
    current_line = add(current_line, "deleted", true);
    netd_conf    = maplist(`line, netd_conf, ``{
	if (lookup(line, "iid", "0") == line_number) {
	    return(current_line);
	}
	else {
	    return(line);
	}
    });
};


/**
 * add a line in DB
 * @param new_line new_line contains new entry for global netd_conf configuration
 * @return void
 */
global define void addLine(map new_line) ``{
    // add
    new_line  = add(new_line, "changed", true);
    netd_conf = add(netd_conf, new_line);
    return;
}

/**
 * Change a line in DB
 * @param new_line new_line contains changes for entry in netd_conf
 * @param line_number line_number contains iid of changed entry in netd_conf
 * @return map If success, returnes updated line
 */
global define changeLine(map new_line, any line_number) ``{
    // entry was changed - so set "changed" flag to true
    new_line  = add(new_line, "changed", true);
    netd_conf = maplist(`line, netd_conf, ``{
	if (lookup(line, "iid", "0") == line_number) {
	    return(new_line);
	}
	else {
	    return(line);
	}
    });
};


/* EOF */
}
