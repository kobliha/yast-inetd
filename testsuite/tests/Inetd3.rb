# encoding: utf-8

# ***************************************************************************
#
# Copyright (c) 2000 - 2012 Novell, Inc.
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of version 2 of the GNU General Public License as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, contact Novell, Inc.
#
# To contact Novell about this file by physical or electronic mail,
# you may find current contact information at www.novell.com
#
# ***************************************************************************
# File:        testsuite/tests/Inetd3.ycp
# Package:     Configuration of inetd
# Summary:     Testsuite for Reading and Writing (x)inetd configuration
# Authors:     Petr Hadraba <phadraba@suse.cz>
#
# $Id$
#
# This is testsuite for Inetd.ycp source file.
# These tests are checking reading and writing
# of inetd or xinetd services.
module Yast
  class Inetd3Client < Client
    def main
      # testedfiles: Inetd.ycp Service.ycp Testsuite.ycp

      Yast.include self, "testsuite.rb"
      Yast.import "Inetd"
      Yast.import "Report"
      Yast.import "Progress"

      Progress.off
      Report.DisplayErrors(false, 0)

      @s = {
        "iid"         => "whatever",
        "comment"     => "My service",
        "enabled"     => true,
        "service"     => "finger",
        "socket_type" => "stream",
        "protocol"    => "tcp",
        "wait"        => false,
        "user"        => "nobody",
        "server"      => "/usr/sbin/tcpd",
        "server_args" => "in.fingerd -w"
      }

      @service_on = { "start" => ["3", "5"], "stop" => ["3", "5"] }

      @service_off = { "start" => [], "stop" => [] }

      @READ = {
        "target"    => { "size" => -1 },
        "init"      => {
          "scripts" => {
            "exists"   => true,
            "runlevel" => { "xinetd" => @service_off },
            "comment"  => { "xinetd" => {} }
          }
        },
        "targetpkg" => { "installed" => true },
        "etc"       => { "xinetd_conf" => { "services" => [@s] } }
      }

      @READXinetd = {}
      @READXinetd = Builtins.eval(@READ)
      Ops.set(
        @READXinetd,
        ["init", "scripts", "runlevel", "xinetd"],
        @service_off
      )
      Ops.set(@READXinetd, ["etc", "inetd_conf"], {})

      @READXinetd2 = {}
      @READXinetd2 = Builtins.eval(@READ)
      Ops.set(
        @READXinetd2,
        ["init", "scripts", "runlevel", "xinetd"],
        @service_on
      )
      Ops.set(@READXinetd2, ["etc", "inetd_conf"], {})

      @WRITE = {}

      @EXECUTE = {
        "target" => {
          "bash_output" => { "exit" => 0, "stdout" => "", "stderr" => "" },
          "mkdir"       => true
        }
      }

      # xinetd service is installed
      # xinetd is stopped
      # User will save and use xinetd
      DUMP("\nRead  --- read all services\n")

      TEST(lambda { Inetd.Read }, [@READXinetd, @WRITE, @EXECUTE], nil)


      DUMP("\nWrite --- write xinetd")
      TEST(lambda { Inetd.Write }, [@READXinetd, @WRITE, @EXECUTE], nil)

      DUMP("\n  ---\n")


      TEST(lambda { Inetd.Read }, [@READXinetd2, @WRITE, @EXECUTE], nil)

      DUMP("\nWrite --- write xinetd")
      TEST(lambda { Inetd.Write }, [@READXinetd2, @WRITE, @EXECUTE], nil)

      DUMP("\n  ---\n")

      nil
    end
  end
end

Yast::Inetd3Client.new.main
