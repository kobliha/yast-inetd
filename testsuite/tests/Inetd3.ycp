/**
 * File:        testsuite/tests/Inetd3.ycp
 * Package:     Configuration of inetd
 * Summary:     Testsuite for Reading and Writing (x)inetd configuration
 * Authors:     Petr Hadraba <phadraba@suse.cz>
 *
 * $Id$
 *
 * This is testsuite for Inetd.ycp source file.
 * These tests are checking reading and writing
 * of inetd or xinetd services.
 */

{

include "testsuite.ycp";
import "Inetd";
import "Report";

Report::DisplayErrors(false, 0);

map s = $[
    "iid": "whatever",
    "comment": "My service",
    "enabled": true,
    "service": "finger",
    "socket_type": "stream",
    "protocol": "tcp",
    "wait": false,
    "user": "nobody",
    "server": "/usr/sbin/tcpd",
    "server_args": "in.fingerd -w",
    ];

map service_on = $[
    "start": [ "3", "5"],
    "stop":  [ "3", "5"],
];

map service_off = $[
    "start": [],
    "stop":  [],
];

map READ = $[
    "target" : $[
	"size" : -1,
    ],
    "init" : $[
	"scripts" : $[
	    "exists" : true,
            "runlevel": $[
                "inetd":  service_off,
                "xinetd": service_off,
            ],
            "comment": $[
                "inetd":  $[],
                "xinetd": $[],
            ],
	]
    ],
    "targetpkg": $[
        "installed": true,
    ],
    "etc" : $[
	"inetd_conf" : $[
	    "services" : [ s ],
	    ],
        "xinetd_conf": $[
            "services":  [ s ],
            ],
	],
];

map READXinetd = $[];
READXinetd =(map) eval(READ);
READXinetd["init", "scripts", "runlevel", "xinetd"] = service_off;
READXinetd["etc", "inetd_conf"]                     = $[ ];

map READXinetd2 = $[];
READXinetd2 =(map) eval(READ);
READXinetd2["init", "scripts", "runlevel", "xinetd"] = service_on;
READXinetd2["etc", "inetd_conf"]                     = $[ ];

map WRITE = $[
];

map EXECUTE = $[
    "target": $[
        "bash_output": $[
            "exit":   0,
            "stdout": "",
            "stderr": "",
        ],
        "mkdir": true,
    ],
];

/*
 * xinetd service is installed
 * xinetd is stopped
 * User will save and use xinetd
 */
DUMP("\nxinetd is stopped and will be started\n");
DUMP("\nRead  --- read all services\n");

// filled by PackageDialog()
Inetd::netd_service = `both;

TEST(``(Inetd::Read()),  [READXinetd, WRITE, EXECUTE], nil);

// filled by InetdDialog()
Inetd::inetd_status     = -1; // not used
Inetd::xinetd_status    = 3;  // unused
Inetd::inetd_installed  = 1;  // not installed
Inetd::xinetd_installed = 0;

// force inetd (by default filled by InetdDialog() according to user choice)
Inetd::netd_conf          = Inetd::xinetd_conf;
Inetd::configured_service = `xinetd;
Inetd::netd_status        = 0; // want to start

DUMP("\nWrite --- write xinetd");
DUMP("  xinetd is installed and stopped\n");
TEST(``(Inetd::Write()), [READXinetd, WRITE, EXECUTE], nil);

DUMP("\n  ---\n");

/*
 * xinetd service is installed
 * xinetd is running
 * User will save and use xinetd
 */
DUMP("\nxinetd is running and will be reconfigured\n");
DUMP("\nRead  --- read all services\n");

// filled by PackageDialog()
Inetd::netd_service = `both;

TEST(``(Inetd::Read()),  [READXinetd2, WRITE, EXECUTE], nil);

// filled by InetdDialog()
Inetd::inetd_status     = -1; // not used
Inetd::xinetd_status    = 0;  // running
Inetd::inetd_installed  = 1;  // not installed
Inetd::xinetd_installed = 0;

// force inetd (by default filled by InetdDialog() according to user choice)
Inetd::netd_conf          = Inetd::xinetd_conf;
Inetd::configured_service = `xinetd;
Inetd::netd_status        = 0; // want to start

DUMP("\nWrite --- write xinetd");
DUMP("  xinetd is installed and running\n");
TEST(``(Inetd::Write()), [READXinetd2, WRITE, EXECUTE], nil);

DUMP("\n  ---\n");

/*
 * xinetd service is installed
 * xinetd is stopped
 * User will stop xinetd
 */
DUMP("\nxinetd is stopped and will be unchanged\n");
DUMP("\nRead  --- read all services\n");

// filled by PackageDialog()
Inetd::netd_service = `both;

TEST(``(Inetd::Read()),  [READXinetd, WRITE, EXECUTE], nil);

// filled by InetdDialog()
Inetd::inetd_status     = -1; // not installed
Inetd::xinetd_status    = 3;  // unused
Inetd::inetd_installed  = 1;  // not installed
Inetd::xinetd_installed = 0;

// force inetd (by default filled by InetdDialog() according to user choice)
Inetd::netd_conf          = Inetd::xinetd_conf;
Inetd::configured_service = `xinetd;
Inetd::netd_status        = -1; // want to stop

DUMP("\nWrite --- write xinetd");
DUMP("  xinetd is installed and stopped\n");
TEST(``(Inetd::Write()), [READXinetd, WRITE, EXECUTE], nil);

DUMP("\n  ---\n");

/*
 * xinetd service is installed
 * xinetd is running
 * User will stop xinetd
 */
DUMP("\nxinetd is running and will be stopped\n");
DUMP("\nRead  --- read all services\n");

// filled by PackageDialog()
Inetd::netd_service = `both;

TEST(``(Inetd::Read()),  [READXinetd2, WRITE, EXECUTE], nil);

// filled by InetdDialog()
Inetd::inetd_status     = -1; // not used
Inetd::xinetd_status    = 0;  // running
Inetd::inetd_installed  = 1;  // not installed
Inetd::xinetd_installed = 0;

// force inetd (by default filled by InetdDialog() according to user choice)
Inetd::netd_conf          = Inetd::xinetd_conf;
Inetd::configured_service = `xinetd;
Inetd::netd_status        = -1; // want to stop

DUMP("\nWrite --- write xinetd");
DUMP("  xinetd is installed and running\n");
TEST(``(Inetd::Write()), [READXinetd2, WRITE, EXECUTE], nil);

DUMP("\n  ---\n");

}
