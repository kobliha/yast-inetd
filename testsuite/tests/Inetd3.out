Dump	
Dump	xinetd is stopped and will be started
Dump	
Dump	
Dump	Read  --- read all services
Dump	
Read	.etc.inetd_conf.services nil
Read	.etc.xinetd_conf.services [$["comment":"My service", "enabled":true, "iid":"whatever", "protocol":"tcp", "server":"/usr/sbin/tcpd", "server_args":"in.fingerd -w", "service":"finger", "socket_type":"stream", "user":"nobody", "wait":false]]
Return	true
Dump	
Dump	Write --- write xinetd
Dump	  xinetd is installed and stopped
Dump	
Write	.etc.xinetd_conf.services [$["comment":"My service", "enabled":true, "iid":"whatever", "protocol":"tcp", "server":"/usr/sbin/tcpd", "server_args":"in.fingerd -w", "service":"finger", "socket_type":"stream", "user":"nobody", "wait":false]] true
Read	.init.scripts.exists "xinetd" true
Execute	.target.bash "/etc/init.d/xinetd start" $["TERM":"raw"] 0
Read	.init.scripts.exists "xinetd" true
Read	.init.scripts.exists "xinetd" true
Read	.init.scripts.exists "xinetd" true
Read	.init.scripts.runlevel "xinetd" $["inetd":$["start":[], "stop":[]], "xinetd":$["start":[], "stop":[]]]
Read	.init.scripts.comment "xinetd" $["inetd":$[], "xinetd":$[]]
Execute	.target.bash_output "/sbin/insserv -d /etc/init.d/xinetd" $["exit":0, "stderr":"", "stdout":""]
Return	true
Dump	
Dump	  ---
Dump	
Dump	
Dump	xinetd is running and will be reconfigured
Dump	
Dump	
Dump	Read  --- read all services
Dump	
Read	.etc.inetd_conf.services nil
Read	.etc.xinetd_conf.services [$["comment":"My service", "enabled":true, "iid":"whatever", "protocol":"tcp", "server":"/usr/sbin/tcpd", "server_args":"in.fingerd -w", "service":"finger", "socket_type":"stream", "user":"nobody", "wait":false]]
Return	true
Dump	
Dump	Write --- write xinetd
Dump	  xinetd is installed and running
Dump	
Write	.etc.xinetd_conf.services [$["comment":"My service", "enabled":true, "iid":"whatever", "protocol":"tcp", "server":"/usr/sbin/tcpd", "server_args":"in.fingerd -w", "service":"finger", "socket_type":"stream", "user":"nobody", "wait":false]] true
Execute	.target.bash "/etc/init.d/xinetd force-reload" $["TERM":"raw"] 0
Return	true
Dump	
Dump	  ---
Dump	
Dump	
Dump	xinetd is stopped and will be unchanged
Dump	
Dump	
Dump	Read  --- read all services
Dump	
Read	.etc.inetd_conf.services nil
Read	.etc.xinetd_conf.services [$["comment":"My service", "enabled":true, "iid":"whatever", "protocol":"tcp", "server":"/usr/sbin/tcpd", "server_args":"in.fingerd -w", "service":"finger", "socket_type":"stream", "user":"nobody", "wait":false]]
Return	true
Dump	
Dump	Write --- write xinetd
Dump	  xinetd is installed and stopped
Dump	
Write	.etc.xinetd_conf.services [$["comment":"My service", "enabled":true, "iid":"whatever", "protocol":"tcp", "server":"/usr/sbin/tcpd", "server_args":"in.fingerd -w", "service":"finger", "socket_type":"stream", "user":"nobody", "wait":false]] true
Return	true
Dump	
Dump	  ---
Dump	
Dump	
Dump	xinetd is running and will be stopped
Dump	
Dump	
Dump	Read  --- read all services
Dump	
Read	.etc.inetd_conf.services nil
Read	.etc.xinetd_conf.services [$["comment":"My service", "enabled":true, "iid":"whatever", "protocol":"tcp", "server":"/usr/sbin/tcpd", "server_args":"in.fingerd -w", "service":"finger", "socket_type":"stream", "user":"nobody", "wait":false]]
Return	true
Dump	
Dump	Write --- write xinetd
Dump	  xinetd is installed and running
Dump	
Write	.etc.xinetd_conf.services [$["comment":"My service", "enabled":true, "iid":"whatever", "protocol":"tcp", "server":"/usr/sbin/tcpd", "server_args":"in.fingerd -w", "service":"finger", "socket_type":"stream", "user":"nobody", "wait":false]] true
Read	.init.scripts.exists "xinetd" true
Execute	.target.bash "/etc/init.d/xinetd stop" $["TERM":"raw"] 0
Read	.init.scripts.exists "xinetd" true
Read	.init.scripts.exists "xinetd" true
Read	.init.scripts.exists "xinetd" true
Read	.init.scripts.runlevel "xinetd" $["inetd":$["start":[], "stop":[]], "xinetd":$["start":["3", "5"], "stop":["3", "5"]]]
Read	.init.scripts.comment "xinetd" $["inetd":$[], "xinetd":$[]]
Execute	.target.bash_output "/sbin/insserv -r /etc/init.d/xinetd" $["exit":0, "stderr":"", "stdout":""]
Return	true
Dump	
Dump	  ---
Dump	
